utils/src/Vulkan/Utils/Initialization.hs

```diff
@@ -1,146 +1,235 @@
-                     (instanceCreateInfo :: InstanceCreateInfo es)
-          }
-    inst <- createInstanceWithExtensions
-      requiredLayers
-      ("VK_LAYER_KHRONOS_validation" : optionalLayers)
-      (EXT_DEBUG_UTILS_EXTENSION_NAME : requiredExtensions)
-      (EXT_VALIDATION_FEATURES_EXTENSION_NAME : optionalExtensions)
-      instanceCreateInfo'
-    _ <- withDebugUtilsMessengerEXT inst
-                                    debugMessengerCreateInfo
-                                    Nothing
-                                    allocate
-    pure inst
-
 -- | Create an 'Instance' with some layers and extensions, the layers and
 -- extensions will be added to the provided 'InstanceCreateInfo'.
 --
+-- Returns all the names of instance layers which were enabled.
+--
 -- Will throw an 'IOError in the case of missing layers or extensions. Details
 -- on missing layers and extensions will be reported in stderr.
-createInstanceWithExtensions
-  :: forall es m
-   . (Extendss InstanceCreateInfo es, PokeChain es, MonadResource m)
-  => [ByteString]
-  -- ^ Required layers
-  -> [ByteString]
-  -- ^ Optional layers
-  -> [ByteString]
-  -- ^ Required extensions
-  -> [ByteString]
-  -- ^ Optional extensions
-  -> InstanceCreateInfo es
-  -> m Instance
-createInstanceWithExtensions requiredLayers optionalLayers requiredExtensions optionalExtensions instanceCreateInfo
+createInstanceFromRequirements
+  :: forall m
+   . MonadResource m
+  => Requirements
+  -> InstanceCreateInfo '[]
+  -> m (Instance, [ByteString])
+createInstanceFromRequirements
+  ( Requirements
+    { version                 = ReqAndOpt reqVersion optVersion
+    , instanceLayers          = ReqAndOpt requiredLayers optionalLayers
+    , instanceExtensions      = reqAndOptExtensions
+    , enableSettingsOtherThan = enableSettingsOtherThan :: HashSet ByteString -> Chain settings
+    }
+  )
+  instanceCreateInfo
   = do
     --
-    -- First get the layers, they're needed to get the list of supported
+    -- We start by checking that the Vulkan API version
+    -- that the user asked for is compatible with the requirements.
+    --
+    let
+      mbGivenVersion :: Maybe Word32
+      mbGivenVersion = ( apiVersion :: ApplicationInfo -> Word32 ) <$> applicationInfo ( instanceCreateInfo :: InstanceCreateInfo '[] )
+    for_ mbGivenVersion $ \ givenVersion -> do
+      unless (givenVersion >= reqVersion)
+        ( liftIO $ noSuchThing
+          ( "Requirements imply a minimum Vulkan API version " <> showVersion reqVersion <> ",\n\
+            \but the version specified by the application is " <> showVersion givenVersion <> "."
+          )
+        )
+      unless (givenVersion >= optVersion)
+        ( sayErr
+          ( "Optional preference: minimum Vulkan API version " <> showVersion optVersion <> ",\n\
+            \but the version specified by the application is " <> showVersion givenVersion <> "."
+          )
+        )
```

Comment:
Yeah, we should use some more structured error type here, can do that later on though.